src/main/webapp/js/instructorFeedbackResults.js

```diff
@@ -205,40 +205,110 @@ function updateStatsCheckBox() {
-        var i = 0;
-        for (var idx = 0; idx < panels.length; idx++) {
-            if ($(panels[idx]).attr('class').indexOf('in') === -1) {
-                setTimeout(showSingleCollapse, 50 * i, panels[idx]);
-                i++;
-            }
-        }
-        var htmlString = $(e).html();
-        htmlString = htmlString.replace(expand, collapse);
-        $(e).html(htmlString);
-        var tooltipString = $(e).attr('data-original-title').replace(expand, collapse);
-        $(e).attr('title', tooltipString).tooltip('fixTitle').tooltip('show');
+/**
+ * Expands or collapses all panels when collapse/expand panels button is clicked.
+ * @param {DOM} element - The element that was clicked to invoke {@code #expandOrCollapsePanels}
+ * @param {DOM} panels - The panels to be expanded/collapsed. not defined if {@param element} is
+ *         collapse panels button.
+ */
+function expandOrCollapsePanels(element, panels) {
+    var STRING_EXPAND = 'Expand';
+    var STRING_COLLAPSE = 'Collapse';
+    var targetPanels = panels || $('div.panel-collapse');
+    var isElementAnExpandButton = $(element).html().trim().startsWith(STRING_EXPAND);
+
+    if (isElementAnExpandButton) {
+        expandPanels(element, targetPanels);
+        replaceInHtml(element, STRING_EXPAND, STRING_COLLAPSE);
     } else {
-        var j = 0;
-        for (var k = 0; k < panels.length; k++) {
-            if ($(panels[k]).attr('class').indexOf('in') !== -1) {
-                setTimeout(hideSingleCollapse, 100 * j, panels[k]);
-                j++;
-            }
+        collapsePanels(targetPanels);
+        replaceInHtml(element, STRING_COLLAPSE, STRING_EXPAND);
+    }
+}
+
+/**
+ * Expands all panels.
+ * @param {DOM} element - The element that was clicked to invoke {@code #expandOrCollapsePanels}
+ */
+function expandPanels(element, panels) {
+    // Expand/collapse buttons on AJAX-loaded panels have a collapse-panels-button class
+    var areAjaxLoadedPanels = $(element).is($('#collapse-panels-button'));
+    var TWENTIETH_SECOND_TO_MILLI_SECOND = 50;
+
+    for (var idx = 0; idx < panels.length; idx++) {
+        expandPanel(panels[idx], idx * TWENTIETH_SECOND_TO_MILLI_SECOND, areAjaxLoadedPanels);
+    }
+}
+
+/**
+ * Expands {@code panel}. If the panel data is not loaded, load it by Ajax first before expansion.
+ * @param {int} timeOut - is in milliseconds
+ */
+function expandPanel(panel, timeOut, isAjaxLoadedPanel) {
+    var isPanelAlreadyExpanded = $(panel).hasClass('in');
+    if (isPanelAlreadyExpanded) {
+        return;
+    }
+
+    if (isAjaxLoadedPanel) {
+        var isElementClickedToLoadAjax = checkAndLoadPanelByAjax(panel);
+        if (isElementClickedToLoadAjax) {
+            // When the element has Ajax class, its panel is in collapsed state. Clicking will expand
+            // its panels.
+            return;
+        }
+    }
+    // expand this panel
+    setTimeout(showSingleCollapse, timeOut, panel);
```

Comment:
That's the problem with doing multiple things in a function. Names such as `foodAndBar` is always not a good sign.
narrow it down to just one job
`getElementToClickForAjaxLoading()`


```js
var $elementToClickForAjaxLoading = getElementToClickForAjaxLoading(panel);

if ($elementToClickForAjaxLoading !== 0) { 
    $elementToClickForAjaxLoading.click();
    // panel is expanded automatically after ajax loading
} else {
    // expand this panel without ajax loading
    setTimeout(showSingleCollapse, timeOut, panel);
}